cmd/agent/app/snmpwalk.go

```diff
@@ -0,0 +1,316 @@
+		if setVersion == gosnmp.Version3 {
+			// Authentication Protocol
+			switch strings.ToLower(authProt) {
+			case "":
+				authProtocol = gosnmp.NoAuth
+			case "md5":
+				authProtocol = gosnmp.MD5
+			case "sha":
+				authProtocol = gosnmp.SHA
+			case "sha224":
+				authProtocol = gosnmp.SHA224
+			case "sha256":
+				authProtocol = gosnmp.SHA256
+			case "sha384":
+				authProtocol = gosnmp.SHA384
+			case "sha512":
+				authProtocol = gosnmp.SHA512
+			default:
+				fmt.Printf("Unsupported authentication protocol: %s \n", authProt)
+				cmd.Help() //nolint:errcheck
+				os.Exit(1)
+				return nil
+			}
+
+			// Privacy Protocol
+			switch strings.ToLower(privProt) {
+			case "":
+				privProtocol = gosnmp.NoPriv
+			case "des":
+				privProtocol = gosnmp.DES
+			case "aes":
+				privProtocol = gosnmp.AES
+			case "aes192":
+				privProtocol = gosnmp.AES192
+			case "aes192c":
+				privProtocol = gosnmp.AES192C
+			case "aes256":
+				privProtocol = gosnmp.AES256
+			case "aes256c":
+				privProtocol = gosnmp.AES256C
+			default:
+				fmt.Printf("Unsupported privacy protocol: %s \n", privProt)
+				cmd.Help() //nolint:errcheck
+				os.Exit(1)
+				return nil
+			}
+
+			// MsgFlags
+			switch strings.ToLower(securityLevel) {
+			case "":
+				msgFlags = gosnmp.NoAuthNoPriv
+				if privKey != "" {
+					msgFlags = gosnmp.AuthPriv
+				} else if authKey != "" {
+					msgFlags = gosnmp.AuthNoPriv
+				}
+
+			case "noauthnopriv":
+				msgFlags = gosnmp.NoAuthNoPriv
+			case "authpriv":
+				msgFlags = gosnmp.AuthPriv
+			case "authnopriv":
+				msgFlags = gosnmp.AuthNoPriv
+			default:
+				fmt.Printf("Unsupported security level: %s \n", securityLevel)
+				cmd.Help() //nolint:errcheck
+				os.Exit(1)
+				return nil
+			}
+		}
+		// Set SNMP parameters
+		snmp = gosnmp.GoSNMP{
+			Target:    deviceIP,
+			Port:      port,
+			Community: communityString,
+			Transport: "udp",
+			Version:   setVersion,
+			Timeout:   time.Duration(timeout * int(time.Second)),
+			Retries:   retries,
+			// v3
+			SecurityModel: gosnmp.UserSecurityModel,
+			ContextName:   snmpContext,
+			MsgFlags:      msgFlags,
+			SecurityParameters: &gosnmp.UsmSecurityParameters{
+				UserName:                 user,
+				AuthenticationProtocol:   authProtocol,
+				AuthenticationPassphrase: authKey,
+				PrivacyProtocol:          privProtocol,
+				PrivacyPassphrase:        privKey,
+			},
+		}
+		// Establish connection
+		err := snmp.Connect()
+		if err != nil {
+			fmt.Printf("Connect err: %v\n", err)
+			os.Exit(1)
+			return nil
+		}
+		defer snmp.Conn.Close()
+
+		// Perform a snmpwalk using Walk for all versions
+		err = snmp.Walk(oid, printValue)
+		if err != nil {
+			fmt.Printf("Walk Error: %v\n", err)
+			os.Exit(1)
+		}
+
+		return nil
+	},
+}
+
+func printValue(pdu gosnmp.SnmpPDU) error {
+
+	fmt.Printf("%s = ", pdu.Name)
+
+	switch pdu.Type {
+	case gosnmp.OctetString:
+		b := pdu.Value.([]byte)
+		fmt.Printf("STRING: %s\n", string(b))
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I think when OctetString contain non-printable chars, we should convert it to Hex String like this:

```
§ snmpwalk -v 2c -c cisco-3850  127.0.0.1:1161 .1.3.6.1.2.1.3.1.1.2.0.1
SNMPv2-SMI::mib-2.3.1.1.2.0.1.192.168.1.1 = Hex-STRING: 6C FB AA 86 E5 63
SNMPv2-SMI::mib-2.3.1.1.2.0.1.192.168.1.2 = Hex-STRING: 6E FB AA 86 E5 63
```

Using current Agent SNMP Walk we are getting:

```
§ ./bin/agent/agent -c dev/dist/datadog.yaml snmp walk -v 2c -C cisco-3850  127.0.0.1:1161 .1.3.6.1.2.1.3.1.1.2.0.1
.1.3.6.1.2.1.3.1.1.2.0.1.192.168.1.1 = STRING: l��c
.1.3.6.1.2.1.3.1.1.2.0.1.192.168.1.2 = STRING: n��c
```

You can probably use something like this : https://github.com/DataDog/datadog-agent/blob/daa9cb2b8aa5148db1e7cb733d3bcaf56b852980/pkg/snmp/gosnmplib/gosnmp_utils.go#L90-L102

If the OctetString contain non-printable chars, convert it to `Hex-STRING`